dotnet/src/Microsoft.AutoGen/AgentChat/Agents/SocietyOfMindAgent.cs

```diff
@@ -0,0 +1,360 @@
+        {
+            _ = message switch
+            {
+                TextMessage textMessage => transcript.AppendLine($"{message.Source}: {textMessage.Content}"),
+                StopMessage stopMessage => transcript.AppendLine($"{message.Source}: {stopMessage.Content}"),
+                HandoffMessage handoffMessage => transcript.AppendLine($"{message.Source}: {handoffMessage.Context}"),
+
+                MultiModalMessage multiModalMessage => AppendMultiModalMessage(multiModalMessage),
+                _ => throw new InvalidOperationException($"Unexpected message type: {message} in {this.GetType().FullName}"),
+            };
+        }
+
+        return transcript;
+
+        StringBuilder AppendMultiModalMessage(MultiModalMessage message)
+        {
+            foreach (MultiModalData part in message.Content)
+            {
+                transcript.Append($"{message.Source}: ");
+
+                if (part.ContentType == MultiModalData.Type.String)
+                {
+                    transcript.AppendLine(((TextContent)part.AIContent).Text);
+                }
+                else if (part.ContentType == MultiModalData.Type.Image)
+                {
+                    transcript.AppendLine("[Image]");
+                }
+                else
+                {
+                    // Best efforts
+                    transcript.AppendLine(part.AIContent.RawRepresentation?.ToString() ?? part.AIContent.ToString());
+                }
+            }
+
+            return transcript;
+        }
+    }
+
+    private ChatStreamFrame ProduceResponse(string result)
+    {
+        Response response = new Response { Message = new TextMessage { Source = this.Name, Content = result } };
+        return new ChatStreamFrame { Type = ChatStreamFrame.FrameType.Response, Response = response };
+    }
+
+    public override async ValueTask<Response> HandleAsync(IEnumerable<ChatMessage> item, CancellationToken cancellationToken)
+    {
+        // In the Python implementation AssistantAgent and SocietyOfMindAgent have different strategies for
+        // reducing on_messages to on_messages_stream. The former returns the first Response as the final
+        // result, while the latter takes the last
+        Response? response = null;
+        await foreach (ChatStreamFrame frame in this.StreamAsync(item, cancellationToken))
+        {
+            if (frame.Type == ChatStreamFrame.FrameType.Response)
+            {
+                response = frame.Response;
+            }
+        }
+
+        return response ?? throw new InvalidOperationException("No response.");
+    }
+
+    public override async IAsyncEnumerable<ChatStreamFrame> StreamAsync(IEnumerable<ChatMessage> item, [EnumeratorCancellation] CancellationToken cancellationToken)
+    {
+        List<ChatMessage> delta = item.ToList();
+        string taskSpecification = this.FormatTask(this.CreateTranscript(delta));
```

Comment:
We may use multi message list as a task for the inner team to avoid creating transcript string